### PR TITLE
Initial implement on Servicer and Storager

### PR DESCRIPTION
**Pull request by aeinrw**
## Related issues
#2 

##  What did I do
 - [x] Implement `Servicer` Interface
 - [x] Implement `Storager` Interface 
 - [x] Pass intergration-test [Github Actions](https://github.com/aeinrw/go-service-minio/actions)

 ## Known issues
* can‘t `Stat` a dir

### CODE REVIEW

**Comment by Xuanwo** [comment]
ping @Prnyself for the first round review.

File: .github/workflows/integration-test.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+name: "Integration Test"
+
+on: [push]
```
Comment (Xuanwo) [inline .github/workflows/integration-test.yml]: We need to test in `pull_request` as well.

File: .github/workflows/integration-test.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        go: [ "1.15", "1.16" ]
+
+    steps:
+      - name: Set up Go 1.x
+        uses: actions/setup-go@v2
+        with:
+          go-version: ${{ matrix.go }}
+
+      - name: Checkout repository
+        uses: actions/checkout@v2
```
Comment (Xuanwo) [inline .github/workflows/integration-test.yml:30]: How about setup a minio in github action directly?

File: .github/workflows/integration-test.yml [hunk 3]
```diff
@@ -0,0 +1,29 @@
+name: "Integration Test"
+
+on: [push]
+
+jobs:
+  integration_test:
+    name: "Integration Test"
+    runs-on: [Linux]
```
Comment (Xuanwo) [inline .github/workflows/integration-test.yml]: We can set up minio in github action as services, so we don't need to run actions on `Linux`, just `ubuntu-latest`.

File: service.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+		return nil, err
+	}
+	return st, nil
+}
+
+func (s *Service) list(ctx context.Context, opt pairServiceList) (sti *StoragerIterator, err error) {
+	input := &storagePageStatus{
+		bufferSize: defaultListStoragerBufferSize,
+	}
+	input.buckets, err = s.service.ListBuckets(ctx)
```
Comment (Xuanwo) [inline service.go]: We should not send API while creating an iterator, please move them to `nextStoragePage`.

File: service.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+		return nil, err
+	}
+	input.total = len(input.buckets)
+	input.remain = input.total
+	return NewStoragerIterator(ctx, s.nextStoragePage, input), nil
+}
+
+func (s *Service) nextStoragePage(ctx context.Context, page *StoragerPage) error {
+	input := page.Status.(*storagePageStatus)
+	if input.remain < input.bufferSize {
```
Comment (Xuanwo) [inline service.go]: Maybe we don't need to handle the buffer size of `StoragerPage`? We can fill them all at once.

File: storage.go [hunk 6]
```diff
@@ -2,35 +2,174 @@ package minio
 
 import (
 	"context"
-	"io"
-
+	"github.com/beyondstorage/go-storage/v4/pairs"
```
Comment (Xuanwo) [inline storage.go]: Please organize import in following style:

- std libs
- 3rd libs
- self libs

File: storage.go [hunk 7]
```diff
@@ -2,35 +2,174 @@ package minio
...
+		o.Mode = ModeRead
+	}
+	o.ID = rp
+	o.Path = path
+	return o
 }
 
 func (s *Storage) delete(ctx context.Context, path string, opt pairStorageDelete) (err error) {
-	panic("not implemented")
+	if opt.HasObjectMode && opt.ObjectMode.IsDir() {
```
Comment (Xuanwo) [inline storage.go:38]: Does minio has native support for dir? If not, we need to check `VirtualDir` flag before handling dir related logic.

Ref: https://github.com/beyondstorage/specs/blob/master/rfcs/109-redesign-features.md

File: storage.go [hunk 8]
```diff
@@ -2,35 +2,174 @@ package minio
...
+	rp := s.getAbsPath(path)
+	err = s.client.RemoveObject(ctx, s.bucket, rp, minio.RemoveObjectOptions{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Storage) deleteDir(ctx context.Context, path string, opt pairStorageDelete) (err error) {
+	objIter, err := s.List(path, pairs.WithListMode(ListModePrefix))
```
Comment (Xuanwo) [inline storage.go]: Call list in `delete` is not allowed, please only remove *one* object.

Ref:

- https://github.com/beyondstorage/specs/blob/master/rfcs/46-idempotent-delete.md
- https://github.com/beyondstorage/specs/blob/5712ecca0ca3a4e1552ea2759404a62f69a04fcf/definitions/operations.toml#L149-L159

File: storage.go [hunk 9]
```diff
@@ -2,35 +2,174 @@ package minio
...
+			rp += "/"
+		}
+	case opt.ListMode.IsPrefix():
+		options.Recursive = true
+	default:
+		return nil, services.ListModeInvalidError{Actual: opt.ListMode}
+	}
+	options.Prefix = rp
+
+	input.objChan = s.client.ListObjects(ctx, s.bucket, options)
```
Comment (Xuanwo) [inline storage.go]: Please don't call API in `list`, move to `nextObjectPage` instead.

File: storage.go [hunk 10]
```diff
@@ -2,35 +2,174 @@ package minio
...
+	meta.Name = s.bucket
+	meta.WorkDir = s.workDir
+	return meta
 }
 
 func (s *Storage) read(ctx context.Context, path string, w io.Writer, opt pairStorageRead) (n int64, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+
+	output, err := s.client.GetObject(ctx, s.bucket, rp, minio.GetObjectOptions{})
```
Comment (Xuanwo) [inline storage.go:107]: Maybe we need to support `offset` and `size`?

It's OK to me to split into another PR.

File: storage.go [hunk 11]
```diff
@@ -2,35 +2,174 @@ package minio
...
+	return io.Copy(w, rc)
 }
 
 func (s *Storage) stat(ctx context.Context, path string, opt pairStorageStat) (o *Object, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+
+	if opt.HasObjectMode && opt.ObjectMode.IsDir() {
+		rp += "/"
+		//TODO
```
Comment (Xuanwo) [inline storage.go]: It's could be better if we specify `todo` what here.

File: storage.go [hunk 12]
```diff
@@ -2,35 +2,174 @@ package minio
...
+	if opt.HasIoCallback {
+		r = iowrap.CallbackReader(r, opt.IoCallback)
+	}
+
+	options := minio.PutObjectOptions{}
+	if opt.HasContentType {
+		options.ContentType = opt.ContentType
+	}
+	if opt.HasContentMd5 {
+		options.SendContentMd5 = true
```
Comment (Xuanwo) [inline storage.go]: `SendContentMd5` will read content and calculate md5 in memory, we need to read `content-md5` provided by user.

> Specify if you'd like to send content-md5 header with PutObject operation. Note that setting this flag will cause higher memory usage because of in-memory md5sum calculation.

Let's communicate with minio community to have a solution here.

File: utils.go [hunk 13]
```diff
@@ -14,14 +43,208 @@ type Storage struct {
...
+	if err != nil {
+		return nil, err
+	}
+
+	var url string
+	var secure bool
+	switch ep.Protocol() {
+	case endpoint.ProtocolHTTP:
+		url, _, _ = ep.HTTP()
+		url = url[7:]
```
Comment (Xuanwo) [inline utils.go]: What does this for? How about constructing them via `host:port`?

File: utils.go [hunk 14]
```diff
@@ -14,14 +43,208 @@ type Storage struct {
...
+	}
+
+	store := &Storage{
+		client:  s.service,
+		bucket:  opt.Name,
+		workDir: "/",
+	}
+
+	if opt.HasWorkDir {
+		if !strings.HasSuffix(opt.WorkDir, "/") {
```
Comment (Xuanwo) [inline utils.go:174]: It's ok for minio to have a workdir that doesn't end with `/`/

File: utils.go [hunk 15]
```diff
@@ -14,14 +43,208 @@ type Storage struct {
...
+}
+
+// getRelPath will get object storage's rel path.
+func (s *Storage) getRelPath(path string) string {
+	prefix := strings.TrimPrefix(s.workDir, "/")
+	return strings.TrimPrefix(path, prefix)
+}
+
+func (s *Storage) formatFileObject(v minio.ObjectInfo) (o *types.Object, err error) {
+	if v.ETag != "" {
```
Comment (Xuanwo) [inline utils.go]: It looks strange to check mode via `ETag`, any reference for this usage?

File: utils.go [hunk 16]
```diff
@@ -14,14 +43,208 @@ type Storage struct {
...
+		//is object
+		o = s.newObject(false)
+		o.Mode |= types.ModeRead
+	} else {
+		//is dir
```
Comment (Xuanwo) [inline utils.go]: ditto. Minio doesn't have native support for dir, please refer to the previous review.

File: utils.go [hunk 17]
```diff
@@ -14,14 +43,208 @@ type Storage struct {
...
+		o.Mode |= types.ModeDir
+	}
+
+	o.ID = v.Key
+	o.Path = s.getRelPath(v.Key)
+	o.SetEtag(v.ETag)
+	o.SetContentLength(v.Size)
+	o.SetContentType(v.ContentType)
+	o.SetLastModified(v.LastModified)
+
```
Comment (Xuanwo) [inline utils.go:245]: Maybe we need to set storage_class here?

File: .github/workflows/integration-test.yml (hunk 2 above)
Comment (aeinrw) [reply .github/workflows/integration-test.yml:30]: Do you mean running a minio docker in action directly？

Comment (Xuanwo) [reply .github/workflows/integration-test.yml:30]: Yep, docs here: https://docs.github.com/en/actions/guides/about-service-containers

File: service.go (hunk 4 above)
Comment (aeinrw) [reply service.go]: `minio.ListBuckets` just gets a slice including all of buckets.

File: service.go (hunk 5 above)
Comment (aeinrw) [reply service.go]: Get it!

File: storage.go (hunk 7 above)
Comment (aeinrw) [reply storage.go:38]: Minio doesn't support to delete a dir, So I call `list` to achieve it.

File: storage.go (hunk 9 above)
Comment (aeinrw) [reply storage.go]: `minio.ListObjects` gets a `<-chan ObjectInfo` which doesn't achieve buffer.

File: storage.go (hunk 11 above)
Comment (aeinrw) [reply storage.go]: Minio can not get the info of dir. It's a `todo` to figure it out.

File: utils.go (hunk 13 above)
Comment (aeinrw) [reply utils.go]: `minio.New` needs the endpoint without protocal label. The field `secure` handles this as an option.

File: utils.go (hunk 16 above)
Comment (aeinrw) [reply utils.go]: `minio.ListObject` will list dir, but stat can not. So I just pass the items without ETag?

File: service.go (hunk 4 above)
Comment (Xuanwo) [reply service.go]: > minio.ListBuckets just gets a slice including all of buckets.

To me, it's an API call that could fail, timeout. And we need to handle them in next page functions instead of in `list`.

File: storage.go (hunk 7 above)
Comment (Xuanwo) [reply storage.go:38]: > Minio doesn't support to delete a dir, So I call `list` to achieve it.

We don't need to do so. `Delete` only needs to delete one object, and leave the user to implement delete dir logic.

File: storage.go (hunk 9 above)
Comment (Xuanwo) [reply storage.go]: > `minio.ListObjects` gets a `<-chan ObjectInfo` which doesn't achieve buffer.

Here is something misunderstanding here.

I don't care about where we create the buffer, but I do care where API call happens. Please don't call API in `list` directly, move them into `next_page` instead. Or the user could get a strange error in `list`.

File: storage.go (hunk 11 above)
Comment (Xuanwo) [reply storage.go]: > Minio can not get the info of dir. It's a `todo` to figure it out.

Got it, please expand them in code directly.

Notes: 

minio doesn't have native support for dir, so we have to simulate them as described in [GSP-109](https://github.com/beyondstorage/specs/blob/master/rfcs/109-redesign-features.md). We need to add `virtual_dir` for this feature, and need to check this flag every time we need to handle the dir object.

---

It's OK to me to remove dir-related logic and split them into another PR.

File: utils.go (hunk 13 above)
Comment (Xuanwo) [reply utils.go]: `url[7:]` looks magic number to me. How about using the host and port that returned by `ep.HTTP()` to construct the addr?

Or leave a comment to make it clear.

Comment (aeinrw) [reply utils.go]: `http://`and `https://`

Comment (Xuanwo) [reply utils.go]: > `http://`and `https://`

Well, as your reviewer, I do understand that does `url[7:]` means. But for other developers, it lacks readability.

Add clear comments here 

```go
// Trim protocol part to get the addr the minio needs.
// "http://exmaple.com:80" => "exmaple.com:80"
url = url[7:]
```

or rewrite into the following style both looks good to me :

```go
_, host, port := ep.HTTP()
addr := fmt.Sprintf("%s:%s", host, port)
```

File: storage.go [hunk 18]
```diff
@@ -3,34 +3,161 @@ package minio
...
 }
 
 func (s *Storage) list(ctx context.Context, path string, opt pairStorageList) (oi *ObjectIterator, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+	options := minio.ListObjectsOptions{WithMetadata: true}
+	switch {
+	case opt.ListMode.IsDir():
+		options.Recursive = false
+		if !strings.HasSuffix(rp, "/") {
```
Comment (Xuanwo) [inline storage.go:57]: Maybe we need to check `virtual_dir` here?

File: storage.go [hunk 19]
```diff
@@ -3,34 +3,161 @@ package minio
...
+	if err != nil {
+		return err
+	}
+	return nil
 }
 
 func (s *Storage) list(ctx context.Context, path string, opt pairStorageList) (oi *ObjectIterator, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+	options := minio.ListObjectsOptions{WithMetadata: true}
```
Comment (Xuanwo) [inline storage.go]: We support [Object Lazy Stat](https://beyondstorage.io/docs/go-storage/internal/object-lazy-stat), which allow the user to get the information only when they really needed.

So we don't need `WithMetadata: true` here.

File: storage.go [hunk 20]
```diff
@@ -3,34 +3,161 @@ package minio
...
+	if input.objChan == nil {
+		input.objChan = s.client.ListObjects(ctx, s.bucket, input.options)
+	}
+
+	for i := 0; i < input.bufferSize; i++ {
+		v, ok := <-input.objChan
+		if !ok {
+			return IterateDone
+		}
+		if v.Err == nil {
```
Comment (Xuanwo) [inline storage.go]: What if `v.Err != nil`?

File: storage.go (hunk 18 above)
Comment (aeinrw) [reply storage.go:57]: _Please excuse me to explain it in Chinese._

这是我对`VirtualDir`的理解：
`VirtualDir`的本意是让不支持目录功能的存储看起来像支持目录一样，用户可以传进来一个pair用来指定是否需要，就像下面一样
```go
store, err := minio.NewStorager(
    ...,
    minio.WithStorageFeatures(minio.StorageFeatures{
	    VirtualDir: true,
    }),
)
```
如果指定为true，那是不是相应的`Create`、`Delete`、`Stat`也要进行支持，那是不是还要去写删除目录及以下文件的逻辑，因为minio原生不支持。
如果指定为false，那对于minio来说，是不是只能用前缀去进行`List`。

另外，在`go-service-oss`的`List`方法中，好像也没有进行`VirtualDir`的检查。

Comment (Xuanwo) [reply storage.go:57]: > 那是不是相应的Create、Delete、Stat也要进行支持

是的。

对 minio 来说，不存在原生的 stat dir 功能，所以模拟的方式就是当用户去 stat abc/ 的时候，如果这个 object 存在，我们返回的 ObjectMode 设置为 `ModeDir` 而不是 `ModeRead`。

> 那是不是还要去写删除目录及以下文件的逻辑，因为minio原生不支持

不是。Delete 的 API 描述中定义的是删除一个 Object，具体来说是删除一个文件或者**一个空目录**。所以 Delete 不需要支持删除一个目录及其以下文件，只需要能够删除一个形如 `abc/` 的 object 即可。递归删除的功能应该由用户自行实现。

>  如果指定为false，那对于minio来说，是不是只能用前缀去进行List。

好问题，确实应该这样。

因为 list 的实现比较早，当初设计的是通过 ListMode 来使用不同的 list 模式，增加 virtual_dir 的时候没有考虑这个问题，所有目前的 service 都没有在 list 的时候考虑这个问题。我创建了一个 issue 来跟踪这个问题： https://github.com/beyondstorage/specs/issues/135

在这个 PR 中，可以先只实现使用 prefix 来 list 的功能。

Comment (aeinrw) [reply storage.go:57]: 在`integration_test`中，传进去的是`ListModeDir`，通过不了集成测试。

Comment (Xuanwo) [reply storage.go:57]: > 在`integration_test`中，传进去的是`ListModeDir`，通过不了集成测试。

OK，那就先参考 s3 的实现吧，我们会在 https://github.com/beyondstorage/specs/issues/135 中彻底解决这个问题

Comment (Xuanwo) [reply storage.go:57]: https://github.com/beyondstorage/go-service-s3/blob/01e370a0f3c0bf3c763b7150bb794b92a7be76cb/storage.go#L269-L270

**Comment by Prnyself** [comment]
Sorry for the delay review, I'll check this next Monday (7.12)

File: storage.go [hunk 21]
```diff
@@ -3,34 +3,162 @@ package minio
...
+	if err != nil {
+		return err
+	}
+	return nil
 }
 
 func (s *Storage) list(ctx context.Context, path string, opt pairStorageList) (oi *ObjectIterator, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+	options := minio.ListObjectsOptions{WithMetadata: false}
```
Comment (Prnyself) [inline storage.go]: We do not need to specify the `WithMetadata` since `false` is the default value, except it indicates some features or issues.

File: storage.go [hunk 22]
```diff
@@ -3,34 +3,162 @@ package minio
...
+	switch {
+	case opt.ListMode.IsDir():
+		options.Recursive = false
```
Comment (Prnyself) [inline storage.go]: ditto

File: storage.go [hunk 23]
```diff
@@ -3,34 +3,162 @@ package minio
...
 
+	"github.com/minio/minio-go/v7"
+
+	ps "github.com/beyondstorage/go-storage/v4/pairs"
+	"github.com/beyondstorage/go-storage/v4/pkg/iowrap"
+	"github.com/beyondstorage/go-storage/v4/services"
 	. "github.com/beyondstorage/go-storage/v4/types"
 )
 
+const defaultListObjectBufferSize = 100
```
Comment (Prnyself) [inline storage.go:16]: Why we declare default buffer size 100 here? Any suggestion or reference here? Or just an experienced value?

File: utils.go [hunk 24]
```diff
@@ -14,14 +43,208 @@ type Storage struct {
...
+	var sm ObjectSystemMetadata
+	sm.StorageClass = v.StorageClass
+	o.SetSystemMetadata(sm)
+
```
Comment (Prnyself) [inline utils.go:245]: Shall we set `UserMetaData` here?

File: utils.go [hunk 25]
```diff
@@ -14,14 +43,208 @@ type Storage struct {
...
+	o.ID = v.Key
+	o.Path = s.getRelPath(v.Key)
+	o.SetEtag(v.ETag)
+	o.SetContentLength(v.Size)
+	o.SetContentType(v.ContentType)
+	o.SetLastModified(v.LastModified)
+
+	var sm ObjectSystemMetadata
+	sm.StorageClass = v.StorageClass
+	o.SetSystemMetadata(sm)
```
Comment (Prnyself) [inline utils.go]:
```suggestion
	o.SetSystemMetadata(ObjectSystemMetadata{
		StorageClass: v.StorageClass,
	})
```

Comment (Prnyself) [reply utils.go]: Remove unnecessary variable `sm`

File: storage.go [hunk 26]
```diff
@@ -3,34 +3,162 @@ package minio
...
+	for i := 0; i < input.bufferSize; i++ {
+		v, ok := <-input.objChan
+		if !ok {
+			return IterateDone
+		}
+		if v.Err != nil {
+			return v.Err
+		}
+		o, err := s.formatFileObject(v)
+		if err != nil && err != services.ErrObjectNotExist {
```
Comment (Prnyself) [inline storage.go]:
```suggestion
		if err != nil && !errors.Is(err, services.ErrObjectNotExist) {
```

File: storage.go [hunk 27]
```diff
@@ -3,34 +3,162 @@ package minio
...
+			return err
+		}
+		if err == services.ErrObjectNotExist {
```
Comment (Prnyself) [inline storage.go]:
```suggestion
		if errors.Is(err, services.ErrObjectNotExist) {
```

File: utils.go (hunk 24 above)
Comment (Xuanwo) [reply utils.go:245]: Negative, `storage-class` belongs to system metadata instead of user metadata.

File: storage.go (hunk 26 above)
Comment (Xuanwo) [reply storage.go]: Why we need to check `ErrObjectNotExist` in `formatFileObject`?

File: utils.go (hunk 24 above)
Comment (Prnyself) [reply utils.go:245]: I mean, minio returns `UserMetadata` in `ObjectInfo`. Shall we set it in `formatFileObject`?

File: storage.go (hunk 26 above)
Comment (Prnyself) [reply storage.go]: It's because `ErrObjectNotExist` was returned in `formatFileObject`. This introduces another question, why check `v.Etag` and return `ErrObjectNotExist` when it's etag is blank?

File: utils.go [hunk 28]
```diff
@@ -14,14 +43,208 @@ type Storage struct {
...
+
+// getRelPath will get object storage's rel path.
+func (s *Storage) getRelPath(path string) string {
+	prefix := strings.TrimPrefix(s.workDir, "/")
+	return strings.TrimPrefix(path, prefix)
+}
+
+func (s *Storage) formatFileObject(v minio.ObjectInfo) (o *types.Object, err error) {
+	if v.ETag == "" {
+		return nil, services.ErrObjectNotExist
```
Comment (Prnyself) [inline utils.go]: Like mentioned [before](https://github.com/beyondstorage/go-service-minio/pull/6/files#r668456350), why we check `v.Etag` here and return `ErrObjectNotExist` when etag is blank?

File: utils.go (hunk 24 above)
Comment (Xuanwo) [reply utils.go:245]: > I mean, minio returns `UserMetadata` in `ObjectInfo`. Shall we set it in `formatFileObject`?

It's OK for me to implement user metadata support in another PR. I prefer sending more PRs instead of adding more features in the same PR.

Comment (Prnyself) [reply utils.go:245]: I agree with that.

File: storage.go (hunk 26 above)
Comment (aeinrw) [reply storage.go]: minio `list` api returns dir without Etag. Here is an example. So I just pass it.

```
Object:
{39e6d200da076f029669ddc6a5b24f95 client.go 2021-07-02 02:13:49.417 +0000 UTC 404  0001-01-01 00:00:00 +0000 UTC map[] map[] map[] {minio 02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4} [] STANDARD false false  <nil>}
Dir:
{ go/ 0001-01-01 00:00:00 +0000 UTC 0  0001-01-01 00:00:00 +0000 UTC map[] map[] map[] { } []  false false  <nil>}
```

File: utils.go (hunk 28 above)
Comment (aeinrw) [reply utils.go]: ditto

File: storage.go (hunk 26 above)
Comment (Xuanwo) [reply storage.go]: It's better to format this object as a dir object instead of returning `ErrObjectNotExist`.

You can set the Object mode to `ModeDir` instead.

File: storage.go (hunk 23 above)
Comment (aeinrw) [reply storage.go:16]: Just an experienced value.

File: storage.go (hunk 12 above)
Comment (Xuanwo) [reply storage.go]: ping @aeinrw , I think the problem is not resolved, so I opened it again.

If we can't set `content-md5` header directly via SDK, we need to remove the option here.`

Comment (aeinrw) [reply storage.go]: Got it！

Comment (aeinrw) [reply storage.go]: Is there any other question？

Comment (Xuanwo) [reply storage.go]: Mostly LGTM after this fix.

**Review by Xuanwo — APPROVED** [review]

**Comment by Xuanwo** [comment]
Congratulations!

**Comment by aeinrw** [comment]
Thank you so much! It means a lot to me.
